Feature/28 form to add new stores

aggiewb opened the pull request.

aggiewb committed:
Merge pull request #36 from aggiewb/staging

Staging

aggiewb committed:
Add html for each store and corresponding styling

aggiewb committed:
Add function to getNewStore() from form inputs

Add event listener on form which calls getNewStore()

aggiewb committed:
Add function addNewStore(), but leave TODO comment

on things to work on next
Move allStoresTotal from function to global scope

aggiewb committed:
Add id to allStoresTotal data cell

Correct typo of namedItem query
Add query of #all-stores-total data cell to be able to update textContent as a new store is added
Remove to do comment

aggiewb committed:
Add min and max constraints to number inputs

aggiewb committed:
Add parseInt() to all numerical form input values

Add newStoreNameInputValidation() for form store name input
Remove superfluous function calls in addNewStore()

aggiewb committed:
Adjust selector properties and add input[type="submit]

css and #warning css for store name input validation

aggiewb committed:
Update form query from using namedInputs() to

querySelector()
Add function toggleTotalRowHighlight() to call when a new store is added
Add toggle of id setting for p element under store name input to display and to hide depending on validity

aggiewb committed:
Add p element for warning to user when store name is not

valid

aggiewb committed:
Change warning class to more verbose name, and add

id selector for a store name warning message
Add class to hide elements such as store name warning message

aggiewb committed:
Update class and attribute types according to match

changes in css

aggiewb committed:
Update id to more verbose name

aggiewb committed:
Add function allInputsCompleted() that disables submit

until all inputs all valid
Add keyup event listener on form to call allInputsCompleted()

aggiewb committed:
Add disabled attribute on submit button

aggiewb committed:
Add logic to reset form inputs and disable submit button

upon store submission

aggiewb committed:
Wrap table contents and new store form in article element

sections

aggiewb committed:
Add display flex to body to be able to have the table and

form displayed side by side
Add margin top and left to h2 to display the form section evenly next to the table
Remove border radius on the form

aggiewb committed:
Add lint script to run using eslint

aggiewb committed:
Fix linter errors

BatemanVO commented on css/style.css:
```diff
@@ -26,4 +27,43 @@ tr:nth-child(even), .highlight{
+    width: 400px;
+    padding: 10px;
+    border: 1px solid #773344;
+}
```
We have the `form` selector declared twice, on 41 and 37. We should combine these styles so it's easier to see everything being applied to a given selector in one spot.

BatemanVO commented on css/style.css:
```diff
@@ -26,4 +27,43 @@ tr:nth-child(even), .highlight{
+    font-weight: bold;
+}
+
+#hide-name-warning-message {
```
A `display: none` class would be better than an `Id`. Something like `.hide`. Toggling Ids is sort of awkward because an element's Id is supposed to be its unique identifier - changing its Id is tantamount to reassigning its unique identifier, which we don't want to do. Classes are better used for CSS styling as a general rule, where Ids are typically useful for getting elements in JavaScript or for web accessibility purposes.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+        event.preventDefault();
+        const form = event.target;
+        const storeName = form.querySelector('#store-name').value;
+        const minHourlyCustomers = parseInt(form.querySelector('#min-hourly-customer').value);
```
If interested, it looks like if you give a `name` attribute to your form `input` elements, you can reference them as properties on `event.target`. E.g., `name="store-name"` can be referenced as `event.target.name`.

Not sure if it's any faster than using `querySelector`, but it might be a little more concise.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+        const storeName = form.querySelector('#store-name').value;
+        const minHourlyCustomers = parseInt(form.querySelector('#min-hourly-customer').value);
+        const maxHourlyCustomers = parseInt(form.querySelector('#max-hourly-customer').value);
+        const avgCookiesPerCustomer = parseInt(form.querySelector('#avg-cookies-per-customer').value);
```
`parseInt` will convert decimals into integers, but all of Pat's provided cookie averages per customer have been decimals. We should look into using `parseFloat` here.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+        cookieStoreLocations.push(newStore);
+        addNewStore(newStore);
+        form.querySelector('input[type=\'submit\']').disabled = true;
+        Array.from(form.querySelectorAll('input')).forEach(input => {
```
You could [look into using the NOT pseudo-selector](https://developer.mozilla.org/en-US/docs/Web/CSS/:not#:~:text=The%20%3Anot()%20CSS%20pseudo,as%20the%20negation%20pseudo%2Dclass.) to reduce the need of an `if` statement, something like `form.querySelectorAll('input:NOT([value="Enter store"])')`.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+        const newStore = new CookieStore(storeName, minHourlyCustomers, maxHourlyCustomers, avgCookiesPerCustomer);
+        cookieStoreLocations.push(newStore);
+        addNewStore(newStore);
+        form.querySelector('input[type=\'submit\']').disabled = true;
```
I'm surprised this query finds an element, I always thought you had to use double-quotes for attributes in selectors. Interesting to know.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+            document.querySelector('tfoot').removeAttribute('class', 'highlight');
+        } else {
+            document.querySelector('tfoot').setAttribute('class', 'highlight');
+        }
```
Although this function looks good, if you want to get fancy, we could reduce it to two lines:
![Capture](https://user-images.githubusercontent.com/20998142/87713392-f76a1800-c75e-11ea-8da6-8d43561ee063.PNG)

It might be crossing into the territory of "clever is not necessarily better than readable", however. What's happening here is using a ternary to decide what method to call, then using [bracket notation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Property_accessors#Bracket_notation) to dynamically access that method on the returned `HTMLElement`.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+            storeNameInput.setAttribute('class', 'name-warning-input');
+            document.querySelector('#hide-name-warning-message').setAttribute('id', 'name-warning-message');
+        } else {
+            storeNameInput.removeAttribute('class', 'name-warning-input');
```
You might want to look into [the classList property](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) that all elements have and make use of the [add](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/add), [remove](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/remove), or maybe even the [toggle](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle) methods. So then it would be something like `storeNameInput.classList.remove('name-warning-input');`.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+    }
+
+    function allInputsCompleted(event){
+        const form = event.target.parentNode;
```
Although referencing `parentNode` is acceptable with current markup, it could prove problematic if a future developer ever wants to modify the form. What if they move the `input` elements into a `fieldset`? Then `parentNode` won't be the `form`.

We should use `document.getElementById` if the `form` has an Id or `document.querySelector` to get the `form` element instead so that we always retrieve the correct value without relying on an exact HTML structure.

BatemanVO commented on js/sales.js:
```diff
@@ -123,6 +123,65 @@
+    function addNewStore(newStore){
+        newStore.render();
+        const storeTotal = newStore.calculateStoreTotal();
+        document.querySelector('tbody').lastChild.appendChild(createDataCell(storeTotal, 'td'));
```
Good use of the `lastChild` property here, I didn't know that was available. It's probably more efficient than querying `tr:last-child`.

BatemanVO commented on package.json:
```diff
@@ -4,7 +4,8 @@
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "lint": "eslint **/*.js"
```
Good to see you getting more familiar with `package.json` and `npm` scripts. Why don't we look into removing the `test` script for now until we add something like `mocha` or `Jest`?

BatemanVO commented on sales.html:
```diff
@@ -7,13 +7,30 @@
-            <tbody></tbody>
-            <tfoot></tfoot>
-        </table>
+        <article>
```
I like the use of semantic tags to try and separate the content, but I'm not sure `<article>` is the correct one. [The MDN documentation for it mentions that it communicates an element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/article):
> is intended to be independently distributable or reusable

The table could probably be argued to possibly be an `article`, but I don't know about the `form`, given that it's tied pretty closely to the `table`.

Maybe we could remove the `<article>` on line 19 and convert the `<h2>` into a [`<legend> element`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/legend) that's a child of the `form`?

aggiewb replied:
Both of these article sections are being used to display flex, and legends seem better suited for a section of a form that needs a label, not for the whole form itself.

BatemanVO commented on sales.html:
```diff
@@ -7,13 +7,30 @@
+            <h2>New Store Input</h2>
+            <form>
+                <label for="store-name">Store Name</label>
+                <input type="text" id="store-name">
```
Good work here not including a self-closing tag or a closing tag at all. It looks like [void elements aren't supposed to have any kind of closer](https://www.w3.org/TR/2011/WD-html-markup-20110113/syntax.html#:~:text=A%20void%20element%20is%20an,param%20%2C%20source%20%2C%20track%20%2C%20wbr#T).

BatemanVO commented on sales.html:
```diff
@@ -7,13 +7,30 @@
+            <form>
+                <label for="store-name">Store Name</label>
+                <input type="text" id="store-name">
+                <p id="hide-name-warning-message">Please enter a valid store name.</p>
```
As mentioned in the CSS, perhaps we could give this an Id of something like `error-message`, then toggle its display using a `hide` class instead of toggling its `Id`.

BatemanVO reviewed (commented):
Looking really good! I haven't had a chance to clone things down and see how it all looks with items sitting side by side, but what you showed the team last looked very clean.

The functions look quite modular, which is good.

Great work!

aggiewb committed:
Combine two instances of form selectors styles into one

aggiewb committed:
Change instances of ids to class and use the classList add(),

remove(), and toggle() methods to set lists of class styles
Remove toggleTotalRowHighlight() as classList.toggle() does all of that work

aggiewb committed:
Update parseInt to parseFloat as the client wants to be

able to input decimals for average cookies per customers
Update input field for average cookies per customers to include a step attribute for decimals

aggiewb committed:
Remove the test script until something like Mocha or Jest

is added

aggiewb committed:
Change article to more generic section html5 element

aggiewb committed:
Use :NOT pseudo selector in query to be able to remove

the if statement
Change all instances of \' to "

BatemanVO reviewed (approved):
Once the `.parentNode` issue is resolved, go ahead and merge this PR to `staging`.

aggiewb committed:
Update form to be queried with document.querySelector()

Remove unused event parameter
Add else block to account for if a user removes a value from an input
Cache submit input to use in if/else block
